src/GeositeFramework/js/Map.js

```diff
@@ -225,23 +225,33 @@ require(['use!Geosite',
                 if (serviceInfo && service.visible && serviceInfo.pluginObject.showServiceLayersInLegend) {
                     service.visibleLayers.sort(function(a, b) { return a - b; });
                     _.each(service.visibleLayers, function(layerId) {
-                        var layer = _.findWhere(service.layerInfos, {id: layerId});
-                        if (!layer) {
-                            return;
-                        }
-
-                        var serviceLegend = getServiceLegend(service);
-                        if (!serviceLegend) {
-                            return;
-                        }
-
-                        var legend = _.findWhere(serviceLegend, {layerId: layerId});
-                        if (isLayerInScale(service, layer)) {
-                            result.push({
-                                service: service,
-                                layer: layer,
-                                legend: legend
-                            });
+                        var layer,
+                            legend;
+
+                        if (_.isUndefined(service.description) === false) {
+                            if (service.description.match(/WMS/i)) {
+                                layer = _.findWhere(service.layerInfos, {name: layerId});
+                                if (!layer) { return; }
+                                legend = layer.legendURL;
+                            } else {
+                                layer = _.findWhere(service.layerInfos, {id: layerId});
+                                if (!layer) { return; }
+                                var serviceLegend = getServiceLegend(service);
+
+                                if (!serviceLegend) {
+                                    return;
+                                }
+
+                                legend = _.findWhere(serviceLegend, {layerId: layerId});
+                            }
+
+                            if (isLayerInScale(service, layer)) {
+                                result.push({
+                                    service: service,
+                                    layer: layer,
+                                    legend: legend
+                                });
+                            }
```

Comment:
Consider splitting this out into separate functions like `if (isWms(service) { ... } else if (isAgs(service)) { ... }` etc.